restapi/stories.go

```diff
@@ -27,35 +27,165 @@ import (
 	if err != nil {
 		return rest.ErrInternal(api.logger, err)
 	}
 
-	return rest.JSONStatusOK(stories)
+	return rest.JSONStatusOK(api.ReturnTaggedStories(stories))
 }
 
 func (api api) ReturnStoryByID(w http.ResponseWriter, r *http.Request) render.Renderer {
 	var story models.Story
 	id := chi.URLParam(r, "storyID")
 
-	err := api.database.Where("is_visible = true").Where("id=?", id).First(&story).Error
+	err := api.database.Preload("Author").Where("id=?", id).First(&story).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
+			return rest.ErrNotFound(fmt.Sprintf("Could not find story with ID %s", id))
+		}
+		return rest.ErrInternal(api.logger, err)
+	}
+	return rest.JSONStatusOK(api.ReturnTaggedStory(story))
+}
+
+func (api api) EditStoryByID(w http.ResponseWriter, r *http.Request) render.Renderer {
+	var story models.Story
+	id := chi.URLParam(r, "storyID")
+
+	err := api.database.Preload("Author").Where("id=?", id).First(&story).Error
 	if err != nil {
 		if err == gorm.ErrRecordNotFound {
 			return rest.ErrNotFound(fmt.Sprintf("Could not find story with ID %s", id))
 		}
 		return rest.ErrInternal(api.logger, err)
 	}
 
-	return rest.JSONStatusOK(story)
+	imageKey, err := convertImageURLToKeyName(story.ImageURL)
+
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, "ImageUrl of existing image is not valid", err)
+	}
+
+	msg, err := api.DeleteImageInS3(imageKey) //delete existing image
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, msg, err)
+	}
+
+	file, h, err := r.FormFile("image")
+
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, "Error getting the image.", err)
+	}
+	imageURL, err := api.uploadImageTos3(file, h.Size, h.Filename)
+	if err != nil {
+		return rest.ErrInvalidRequest(api.logger, "Error uploading the image to s3.", err)
+	}
+	defer file.Close()
+
+	var author models.Author
+	author.FirstName = r.FormValue("author_first_name")
+	author.LastName = r.FormValue("author_last_name")
+	author.OriginCountry = r.FormValue("author_country")
+
+	errAuthor := api.database.First(&author).Error
+	if errAuthor != nil {
+		if errAuthor == gorm.ErrRecordNotFound {
+			country := countries.ByName(author.OriginCountry)
+			if country == countries.Unknown {
+				return rest.ErrInvalidRequest(api.logger, "Unknown origin country", nil)
+			}
+			if err := api.database.Create(&author).Error; err != nil { //Adding author
+				return rest.ErrInternal(api.logger, errAuthor)
+			}
+		} else {
+			return rest.ErrInternal(api.logger, errAuthor)
+		}
+	}
+	author.Bio = r.FormValue("bio")
+	api.database.Save(&author)
+	story.Author = author
+	story.AuthorFirstName = r.FormValue("author_first_name")
+	story.AuthorLastName = r.FormValue("author_last_name")
+	story.AuthorCountry = r.FormValue("author_country")
+	story.ImageURL = imageURL
+	story.Title = r.FormValue("title")
+	story.Content = r.FormValue("content")
+	story.CurrentCity = r.FormValue("current_city")
```

Comment:
Can we just do a single story instantiation?

Like
```go
story := {
  Author: author,
  AuthorFirstName: r.FormValue("author_first_name")
  ...
}
```

Generally I wanna see this function organized like this:

form data |> processed form data |> story struct |> insert into db

That way we can easily extract one of these steps into a pure function for unit testing or to abstract so that other functions can use it.